block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillGrpcClient.java

```diff
@@ -280,6 +181,170 @@ public void resetStatus() {
+            List<LongRange> intersections = new ArrayList<>();
+            for (LongRange range : ranges) {
+                long intersectionStart = Math.max(targetRange.start(), range.start());
+                long intersectionEnd = Math.min(targetRange.end(), range.end());
+                if (intersectionStart <= intersectionEnd) {
+                    intersections.add(new LongRange(intersectionStart, intersectionEnd));
+                }
+            }
+
+            if (!intersections.isEmpty()) {
+                availability.put(node, intersections);
+                nodeStatusMap.put(node, Status.AVAILABLE);
+            } else {
+                nodeStatusMap.put(node, Status.UNAVAILABLE);
+            }
+        }
+
+        return availability;
+    }
+
+    /**
+     * Selects the best node and chunk to fetch next, based on pre-computed availability.
+     *
+     * @param startBlock the next block number to fetch
+     * @param gapEnd     inclusive end of the gap
+     * @param availability map of node -> available range intersecting the gap
+     * @return optional NodeSelection describing which node to hit and what range to request
+     */
+    public Optional<NodeSelection> selectNextChunk(
+            long startBlock, long gapEnd, @NonNull Map<BackfillSourceConfig, List<LongRange>> availability) {
+        if (startBlock > gapEnd) {
+            return Optional.empty();
+        }
+
+        OptionalLong earliestAvailableStart = findEarliestAvailableStart(startBlock, gapEnd, availability);
+        if (earliestAvailableStart.isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<NodeSelection> candidates =
+                candidatesForEarliest(startBlock, gapEnd, earliestAvailableStart.getAsLong(), availability);
+        return chooseBestCandidate(candidates);
+    }
+
+    /** Locate the earliest start index across all available ranges that can satisfy the request bounds. */
+    private OptionalLong findEarliestAvailableStart(
+            long startBlock, long gapEnd, @NonNull Map<BackfillSourceConfig, List<LongRange>> availability) {
+        long earliestAvailableStart = Long.MAX_VALUE;
+        for (Map.Entry<BackfillSourceConfig, List<LongRange>> entry : availability.entrySet()) {
+            for (LongRange availableRange : entry.getValue()) {
+                long candidateStart = Math.max(startBlock, availableRange.start());
+                if (candidateStart > availableRange.end() || candidateStart > gapEnd) {
+                    continue;
+                }
+                earliestAvailableStart = Math.min(earliestAvailableStart, candidateStart);
+            }
+        }
+
+        if (earliestAvailableStart == Long.MAX_VALUE) {
+            return OptionalLong.empty();
+        }
+
+        return OptionalLong.of(earliestAvailableStart);
+    }
+
+    /** Build the list of nodes that can serve the earliest available start. */
+    private List<NodeSelection> candidatesForEarliest(
```

Comment:
In the future we could optimize this by doing it as part of getting availability.
Right now it's not as costly but something to note as it would simplify the logic